File: androidApp/src/main/java/com/amannegi/kmmdemo/android/ProductDetailsScreen.kt
package com.amannegi.kmmdemo.android

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.navigation.NavController
import androidx.navigation.compose.rememberNavController
import coil.compose.AsyncImage
import com.amannegi.kmmdemo.Product
import com.google.accompanist.pager.ExperimentalPagerApi
import com.google.accompanist.pager.HorizontalPager
import com.google.accompanist.pager.HorizontalPagerIndicator
import com.google.accompanist.pager.rememberPagerState

@OptIn(ExperimentalPagerApi::class)
@Composable
fun ProductDetailsScreen(navController: NavController, product: Product) {
    val scaffoldState = rememberScaffoldState()

    Scaffold(
        scaffoldState = scaffoldState,
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = product.title ?: "Product",
                        style = MaterialTheme.typography.h6
                    )
                },
                navigationIcon = {
                    IconButton(onClick = { navController.popBackStack() }) {
                        Icon(imageVector = Icons.Filled.ArrowBack, contentDescription = "back")
                    }
                }
            )
        }
    ) { scaffoldPadding ->

        Column(
            modifier = Modifier
                .padding(scaffoldPadding)
                .verticalScroll(rememberScrollState())
        ) {
            val pageState = rememberPagerState()

            HorizontalPager(
                state = pageState,
                count = product.images?.size ?: 0,
//                contentPadding = PaddingValues(horizontal = 32.dp)
            ) { page ->
                AsyncImage(
                    model = product.images?.get(page),
                    contentDescription = "Product image $page",
                    contentScale = ContentScale.FillBounds,
                    placeholder = painterResource(id = R.drawable.ic_round_imagesearch_roller_24),
                    modifier = Modifier
                        .fillMaxWidth()
                        .aspectRatio(1f)
                )
            }

            HorizontalPagerIndicator(
                pagerState = pageState,
                modifier = Modifier
                    .align(Alignment.CenterHorizontally)
                    .padding(16.dp)
            )

            Text(
                text = "${product.description}",
                style = MaterialTheme.typography.body1,
                modifier = Modifier
                    .padding(vertical = 8.dp, horizontal = 16.dp)
                    .fillMaxWidth()
            )

            Row(
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                modifier = Modifier.padding(horizontal = 16.dp)
            ) {
                product.discountPercentage?.let {
                    if (product.discountPercentage != 0.0) {
                        Text(
                            text = "${product.discountPercentage}% off",
                            style = MaterialTheme.typography.h5,
                            fontWeight = FontWeight.Bold,
                            color = Color(0xFF02b502)
                        )
                    }
                }
                Text(
                    text = "₹${product.price}",
                    style = MaterialTheme.typography.h5,
                    fontWeight = FontWeight.Bold
                )
            }

            Row(
                modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
                verticalAlignment = Alignment.Bottom
            ) {
                RatingBar(
                    modifier = Modifier.fillMaxWidth(0.3f),
                    numberOfStars = 5,
                    rating = (product.rating ?: 0).toFloat(),
                    starSpacing = 2.dp
                )
                Text(
                    text = "${product.rating}",
                    style = MaterialTheme.typography.caption
                )
            }

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 8.dp)
                    .clip(RoundedCornerShape(8.dp))
                    .background(Color.Gray)
            ) {
                Text(
                    text = "Only ${product.stock} left!",
                    style = MaterialTheme.typography.subtitle1,
                    modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp, vertical = 8.dp),
                    textAlign = TextAlign.Center,
                    color = Color.White
                )
            }

        }
    }
}

@Preview
@Composable
fun ProductDetailsScreenPreview() {
    MyApplicationTheme {
        ProductDetailsScreen(navController = rememberNavController(), product = dummyProduct)
    }
}

File: iosApp/iosApp/Screens/ProductDetailsScreen.swift
//
//  ProductDetailsScreen.swift
//  iosApp
//
//  Created by Aman Negi on 08/10/22.
//  Copyright © 2022 iamanegi. All rights reserved.
//

import SwiftUI
import shared

struct ProductDetailsScreen: View {
    let product: Product
    
    var body: some View {
        ScrollView(.vertical, showsIndicators: false) {
            VStack(alignment: .leading, spacing: 8) {
                
                HorizontalPager(pageCount: (product.images?.count) ?? 0) { index in
                    AsyncImage(
                        url: URL(string: (product.images?[index] ?? ""))) { image in
                            image.resizable()
                                .aspectRatio(1, contentMode: .fit)
                        } placeholder: {
                            Image("placeholder")
                                .resizable()
                                .aspectRatio(1, contentMode: .fit)
                        }
                }
                
                Text(product.description_ ?? "")
                    .font(.system(.body))
                    .padding(.horizontal)
                
                HStack (alignment: .center, spacing: 16) {
                    if let discount = product.discountPercentage {
                        Text("\(discount)% off")
                            .font(.system(.body))
                            .fontWeight(.bold)
                            .foregroundColor(.green)
                    }
                    
                    Text("₹\(product.price ?? 0)")
                        .font(.system(.body))
                        .fontWeight(.bold)
                }.padding(.horizontal)
                
                HStack (alignment: .bottom, spacing: 8) {
            
                    RatingBar(
                        rating: Decimal(floatLiteral: Double(truncating: product.rating!))
                    ).frame(width: UIScreen.main.bounds.width * 0.3)
                        .padding(.vertical)
                    
                    Text("\(product.rating ?? 0)")
                        .font(.system(.caption))
                        .padding(.vertical, 4)

                }.padding(.horizontal)
                
                Text("Only \(product.stock!) left!")
                    .font(.system(.subheadline))
                    .foregroundColor(.white)
                    .padding(.vertical, 8)
                    .frame(maxWidth: .infinity)
                    .background(.gray)
                    .cornerRadius(8)
                    .padding(.horizontal)
                    
            }
        }.navigationTitle(product.title ?? "Product Detail")

    }
}

struct HorizontalPager<Content: View>: View {
    
    let pageCount: Int
    let content: (Int) -> Content
    
    var body: some View {
        LazyHStack {
            TabView {
                ForEach(0..<pageCount, id: \.self) { i in
                    content(i)
                }
            }.aspectRatio(1, contentMode: .fit)
                .frame(width: UIScreen.main.bounds.width)
                .tabViewStyle(PageTabViewStyle())
            
        }
    }
}

struct ProductDetailsScreen_Previews: PreviewProvider {
    static var previews: some View {
        ProductDetailsScreen(product: dummyProduct)
    }
}
